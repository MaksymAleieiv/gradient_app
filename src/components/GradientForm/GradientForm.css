.gradient-form {
    width: min(320px, 90%);
    height: 320px;
    margin: auto;
}

.gradient-form__inner {
    display: flex;
    justify-content: space-between;
}

.gradient-form-input-holder {
    width: 48%;
}

.gradient-form-input {
    width: 100%;
    height: 32px;
    padding-left: 4px;
    font-size: 24px;
    color: var(--main-color);
    background-color: var(--whitey);
    border:none;
}

.gradient-form-input:focus {
    outline: none;
    border:none;
}

.gradient-form-error {
    color: #f95353;
    margin-top: 4px;
    height: 18px;
}

.gradient-form-submit {
    width: 100%;
    height: 32px;
    color: var(--main-color);
    border: none;
    font-size: 22px;
    background: var(--whitey);
    margin-top: 20px;
    cursor: pointer;
    border-radius: var(--border-rad);
}

@media (max-width: 420px) {
    .gradient-form__inner {
        display: block;
    }
    .gradient-form-input-holder {
        width: 100%;
    }
}
