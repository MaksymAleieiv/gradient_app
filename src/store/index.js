import { createStore } from 'redux'
import { gradientsList_Reducer as gradientsList } from './reducers/gradients-list/gradientsList_Reducer'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'

const persistConfig = {
    key: 'root',
    storage
}

const persistedReducer = persistReducer(persistConfig, gradientsList)


let store = createStore(persistedReducer)
let persistor = persistStore(store)
export { store, persistor }
