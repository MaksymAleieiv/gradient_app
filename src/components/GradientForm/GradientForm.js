import { Field, Form, Formik } from 'formik'
import React from 'react'
import { useDispatch } from 'react-redux'
import { useNavigate } from 'react-router'
import { gradientsList_ActionCreators } from '../../store/reducers/gradients-list/gradientsList_ActionCreators'
import './GradientForm.css'

function GradientForm({gradient = {}}) {
    const dispatch = useDispatch();
    const navigate = useNavigate();

    const onSubmit = (values) => {
        if(!!gradient.id) {
            dispatch(gradientsList_ActionCreators.editGradient({...values, id: gradient.id}))
        }
        else {
            dispatch(gradientsList_ActionCreators.addGradient({...values, id: Date.now().toString()}))
        }
        navigate('/')
    }

    const validateHex = (value) => {
        if(!/^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/.test(value)) return 'Not HEX';
        else return;
    }

    return (
        <div data-testid='form'>
            <Formik
                initialValues={{
                    firstColor: gradient.firstColor || '',
                    secondColor: gradient.secondColor || ''
                }}
                onSubmit={onSubmit}
            >
                {({errors, touched, isValid, dirty}) => (
                    <Form className='gradient-form'>
                        <div className='gradient-form__inner'>
                            <div className='gradient-form-input-holder'>
                                <Field name='firstColor' validate={validateHex} className={`gradient-form-input ${touched.firstColor && errors.firstColor && 'error'}`}/>
                                <div className='gradient-form-error'>{touched.firstColor && errors.firstColor}</div>
                            </div>
                            <div className='gradient-form-input-holder'>
                                <Field name='secondColor' validate={validateHex} className={`gradient-form-input ${touched.secondColor && errors.secondColor && 'error'}`}/>
                                <div className='gradient-form-error'>{touched.secondColor && errors.secondColor}</div>
                            </div>
                        </div>
                        {!!gradient.id ? (
                            <button type='submit' disabled={!isValid} className={(isValid) ? 'gradient-form-submit' : 'gradient-form-submit error'}> {gradient.firstColor ? 'Save Changes' : 'Add Gradient'}</button>
                        ) : (
                            <button type='submit' disabled={!(isValid && dirty)} className={(isValid && dirty) ? 'gradient-form-submit' : 'gradient-form-submit error'}> {gradient.firstColor ? 'Save Changes' : 'Add Gradient'}</button>
                        )}
                        
                    </Form>
                )}
            </Formik>
        </div>
    )
}

export default GradientForm
