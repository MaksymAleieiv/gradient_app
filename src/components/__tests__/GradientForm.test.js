import { render, screen, cleanup } from '@testing-library/react';
import {BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer'

import { Provider } from 'react-redux';
import { store } from '../../store/index'
import GradientForm from '../GradientForm/GradientForm';

afterEach(() => {
    cleanup();
})

test('gradient form create snapshot, i want to die', () => {
    const tree = renderer.create(
        <Provider store={store}>
            <Router>
                <GradientForm/>
            </Router>
        </Provider>
            ).toJSON();
    expect(tree).toMatchSnapshot();
})

test('gradient form edit snapshot, i want to die', () => {
    const gradient = { id: '10', firstColor: '#fff', secondColor: '#000' };
    const tree = renderer.create(
        <Provider store={store}>
            <Router>
                <GradientForm gradient={gradient}/>
            </Router>
        </Provider>
            ).toJSON();
    expect(tree).toMatchSnapshot();
})

test('create new gradient form render, initially button should be disabled, i want to die', () => {
    render(
        <Provider store={store}>
            <Router>
                <GradientForm/>
            </Router>
        </Provider>
    )
    const formEl = screen.getByTestId('form');
    expect(formEl).toBeInTheDocument();
    expect(formEl).toHaveTextContent('Add Gradient');
    expect(formEl).toContainHTML('<div data-testid="form"><form action="#" class="gradient-form"><div class="gradient-form__inner"><div class="gradient-form-input-holder"><input name="firstColor" class="gradient-form-input undefined" value=""><div class="gradient-form-error"></div></div><div class="gradient-form-input-holder"><input name="secondColor" class="gradient-form-input undefined" value=""><div class="gradient-form-error"></div></div></div><button type="submit" disabled="" class="gradient-form-submit error"> Add Gradient</button></form></div>');
})

test('edit existing gradient form render, initially button should not be disabled, i want to die', () => {
    const gradient = { id: '1638989386828', firstColor: '#321', secondColor: '#123' };
    render(
        <Provider store={store}>
            <Router>
                <GradientForm gradient={gradient}/>
            </Router>
        </Provider>
    )
    const formEl = screen.getByTestId('form');
    expect(formEl).toBeInTheDocument();
    expect(formEl).toHaveTextContent('Save Changes');
    expect(formEl).toContainHTML('<div data-testid="form"><form action="#" class="gradient-form"><div class="gradient-form__inner"><div class="gradient-form-input-holder"><input name="firstColor" class="gradient-form-input undefined" value="#321"><div class="gradient-form-error"></div></div><div class="gradient-form-input-holder"><input name="secondColor" class="gradient-form-input undefined" value="#123"><div class="gradient-form-error"></div></div></div><button type="submit" class="gradient-form-submit"> Save Changes</button></form></div>');
})