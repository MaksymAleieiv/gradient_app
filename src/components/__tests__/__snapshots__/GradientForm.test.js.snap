// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gradient form create snapshot, i want to die 1`] = `
<div
  data-testid="form"
>
  <form
    action="#"
    className="gradient-form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="gradient-form__inner"
    >
      <div
        className="gradient-form-input-holder"
      >
        <input
          className="gradient-form-input undefined"
          name="firstColor"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        />
        <div
          className="gradient-form-error"
        />
      </div>
      <div
        className="gradient-form-input-holder"
      >
        <input
          className="gradient-form-input undefined"
          name="secondColor"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        />
        <div
          className="gradient-form-error"
        />
      </div>
    </div>
    <button
      className="gradient-form-submit error"
      disabled={true}
      type="submit"
    >
       
      Add Gradient
    </button>
  </form>
</div>
`;

exports[`gradient form edit snapshot, i want to die 1`] = `
<div
  data-testid="form"
>
  <form
    action="#"
    className="gradient-form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="gradient-form__inner"
    >
      <div
        className="gradient-form-input-holder"
      >
        <input
          className="gradient-form-input undefined"
          name="firstColor"
          onBlur={[Function]}
          onChange={[Function]}
          value="#fff"
        />
        <div
          className="gradient-form-error"
        />
      </div>
      <div
        className="gradient-form-input-holder"
      >
        <input
          className="gradient-form-input undefined"
          name="secondColor"
          onBlur={[Function]}
          onChange={[Function]}
          value="#000"
        />
        <div
          className="gradient-form-error"
        />
      </div>
    </div>
    <button
      className="gradient-form-submit"
      disabled={false}
      type="submit"
    >
       
      Save Changes
    </button>
  </form>
</div>
`;
